packages/fxa-settings/src/pages/Signin/SigninTotpCode/container.tsx

```diff
@@ -0,0 +1,111 @@
+
+import { RouteComponentProps, useNavigate, useLocation } from '@reach/router';
+import { useValidatedQueryParams } from '../../../lib/hooks/useValidate';
+import { SigninQueryParams } from '../../../models/pages/signin';
+import { SigninTotpCode } from './index';
+import { gql, useMutation } from '@apollo/client';
+import GleanMetrics from '../../../lib/glean';
+import { logViewEvent } from '../../../lib/metrics';
+import { hardNavigate } from 'fxa-react/lib/utils';
+import { AuthUiErrors } from '../../../lib/auth-errors/auth-errors';
+import { currentAccount, sessionToken } from '../../../lib/cache';
+import { MozServices } from '../../../lib/types';
+import VerificationMethods from '../../../constants/verification-methods';
+
+export const viewName = 'signin-totp-code';
+
+export type SigninTotpCodeContainerProps = {
+  serviceName: MozServices;
+};
+
+export const SignInTotpCodeContainer = ({
+  serviceName,
+}: SigninTotpCodeContainerProps & RouteComponentProps) => {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const storedLocalAccount = currentAccount();
+
+  // TBD: A bit of a work around. In the previous signup step
+  // a new session token should have been generated. Setting
+  // it here to ensure that the authorization header is set
+  // when it comes time to verify the totp code.
+  if (storedLocalAccount?.sessionToken) {
+    sessionToken(storedLocalAccount?.sessionToken);
+  }
+
+  const { queryParamModel } = useValidatedQueryParams(SigninQueryParams);
+  const [verifyTotpCode] = useMutation(gql`
+    mutation VerifyTotp($input: VerifyTotpInput!) {
+      verifyTotp(input: $input) {
+        success
+      }
+    }
+  `);
+
+  const handleNavigation = () => {
+    console.log('!!! handleNavigation', queryParamModel);
+
+    // Note: `redirectTo` looks similar to `redirectPathname`, but they
+    // work slightly differently: `redirectTo` redirects automatically,
+    // while `redirectPathname` redirects after a form submission.
+    if (queryParamModel.redirectToUrl) {
+      hardNavigate(queryParamModel.redirectToUrl);
+    } else if (queryParamModel.redirectPathname) {
+      navigate(queryParamModel.redirectPathname + location.search);
+    } else if (storedLocalAccount?.verificationReason === 'change_password') {
+      navigate(`/post_verify/password/force_password_change${location.search}`);
+    } else {
+      navigate('/settings');
+    }
+  };
+
+  const submitTotpCode = async (code: string) => {
+    try {
+      const result = await verifyTotpCode({
+        variables: {
+          input: {
+            code,
+            service: queryParamModel.service,
+          },
+        },
+      });
+      GleanMetrics.loginConfirmation.submit();
+
+      // Check authentication code
+      if (result.data?.verifyTotp.success === true) {
+        logViewEvent('flow', `${viewName}.submit`);
+        return true;
+      }
+
+      return false;
+    } catch (error) {
+      if (error.errno) {
```

Comment:
If the error is a graphQL error, we likely need to check for an error number in the extensions or we'll always get an unexpected error. Also need to account for throttled errors with a retryAfterLocalized value.

I pulled this out into a utility function: https://github.com/mozilla/fxa/pull/16376/files#diff-6c6342b35ed2e6047d935bf8804fd309cf03b887179cf6cbe34863a8407c8708

Reply:
Thanks!